fw/mmap_buffer.h

```diff
@@ -0,0 +1,237 @@
+
+#define TFW_MMAP_BUFFER_MAX_NAME_LEN 32
+
+#define TFW_MMAP_BUFFER_FULL_SIZE(size) (TFW_MMAP_BUFFER_DATA_OFFSET + size * 2)
+
+/**
+ * @head	- Head offset where the next data write will happen;
+ * @tail	- tail offset where the next data read will happen;
+ * @size	- size of the ring buffer data in bytes;
+ * @mask	- limits head and tail to the buffer size (a power of two), replacing
+ *		  mod for faster indexing;
+ * @cpu		- ID of the CPU tied to this buffer;
+ * @is_ready	- indicates that the buffer is mapped to user space and ready
+ *		  both for writing and reading. Resetting this field signals to
+ *		  user space that it should stop reading, unmap and close the file;
+ * @data	- points to the data start.
+ */
+typedef struct {
+	u64		head;
+	u64		tail;
+	u32		size;
+	u32		mask;
+	u32		cpu;
+#ifdef __KERNEL__
+	atomic_t	is_ready;
+#else
+	int		is_ready;
+#endif
+	char __attribute__((aligned(TFW_MMAP_BUFFER_DATA_OFFSET))) data[];
+} TfwMmapBuffer;
+
+#ifdef __KERNEL__
+
+/**
+ * @buf_page	- Pages allocated for buffer metadata;
+ * @data_page	- pages allocated for data storage.
+ */
+typedef struct {
+	struct page	*buf_page;
+	struct page	*data_page;
+} TfwMMapBufferMem;
+
+/**
+ * @buf			- Per CPU pointers to store pointers to buffers;
+ * @dev_name		- name of the device in /dev;
+ * @size		- size of the memory allocated to every buffer;
+ * @dev_is_opened	- indicates that freeing process started, it's
+ *			  necessary to exclude repeated file opening;
+ * @is_freeing		- indicates that freeing process started, It's
+ *			  necessary to exclude repeated file opening;
+ * @dev			- device structure of the device in /dev;
+ * @dev_major		- the major number of the device in /dev;
+ * @dev_class		- the class of the device;
+ * @mem			- array of structures descripting allocated memory.
+ */
+typedef struct {
+	TfwMmapBuffer __percpu	**buf;
+	char			dev_name[TFW_MMAP_BUFFER_MAX_NAME_LEN];
+	unsigned int		size;
+	atomic_t		dev_is_opened;
+	atomic_t		is_freeing;
+	struct device		*dev;
+	int			dev_major;
+	struct class		*dev_class;
+	TfwMMapBufferMem	mem[];
+} TfwMmapBufferHolder;
+
+/*
+ * The function 'tfw_mmap_buffer_get_room()' returns the size of the memory
+ * region available for writing. Internal state of the buffer (i.e., head or
+ * tail positions) is not modified at this time. As a result, the writing
+ * process can be interrupted at any time, and this function can be called
+ * again to request space for another element without affecting previous calls.
+ *
+ * Once the data has been successfully written, 'tfw_mmap_buffer_commit()' must
+ * be called, passing the actual size of the written data. This function
+ * updates the buffer's internal state to reflect the new data and make the
+ * written space unavailable for further writing.
+ */
+static __always_inline unsigned int
+tfw_mmap_buffer_get_room(TfwMmapBufferHolder *holder, char **data)
+{
+	TfwMmapBuffer *buf = *this_cpu_ptr(holder->buf);
+	u64 size;
```

Comment:
unused

Reply:
More precisely it is set on line 211 and unused later

Reply:
Thanks.
Fixed.